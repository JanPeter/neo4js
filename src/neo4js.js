import { v1 as neo4j } from 'neo4j-driver';
import Utils from './Utils';
import uuid from 'node-uuid';
import Debug from 'debug';
const debug = Debug('neo4js');

import * as neo4jsErrors from './errors';
import Model from './model';
import ModelManager from './model-manager';
import ModelObject from './model-object';
import Query from './query';
import { getSchemaInfo } from './schema-info';

class Neo4js {
  /**
   * Constructor to create a new instance of the Neo4js class
   * @param {String}  username The username which is used to authenticate against the database
   * @param {String}  password The password which is used to authenticate against the database
   * @param {Object}  [options={}] An object with options
   * @param {String}  [options.url=localhost] Bolt url
   * @param {Integer} [options.port=7687] Bolt port
   * @param {Object}  [options.rest={}]
   * @param {String}  [options.rest.url=localhost]
   * @param {Integer} [options.rest.port=7474]
   * @param {Integer} [options.connectionPoolSize=50]
   */
  constructor(username, password, options) {
    this.auth = { username, password };

    const defaultOptions = {
      url: `bolt://localhost`,
      port: 7687,
      rest: {
        url: `http://localhost`,
        port: 7474,
      },
      driver: {
        connectionPoolSize: 50,
      },
    };

    options = Utils._.extend({}, defaultOptions, options || {});
    options.rest = Utils._.extend({}, defaultOptions.rest, options.rest);
    options.driver = Utils._.extend({}, defaultOptions.driver, options.driver);

    // TODO: Maybe add here some more checks...
    const url = `${options.url}:${options.port}`;
    this.driver = neo4j.driver(url, neo4j.auth.basic(username, password), options.driver);

    this.options = options;

    this.models = {};
    this.modelManager = new ModelManager(this);
  }

  /**
   * @param {String | String[]} labels
   * @param {Object} schema
   * @param {Object} schema.{propertyName}
   * @param {Boolean} schema.{propertyName}.index
   * @param {Boolean} schema.{propertyName}.unique
   * @param {Boolean} schema.{propertyName}.exists
   * @param {Function} schema.{propertyName}.defaultValue
   */
  define(labels, schema, hooks) {
    const model = new Model(labels, schema, hooks, this);
    this.modelManager.add(model);
    return model;
  }

  /**
   * @param {String | String[]} labels
   */
  getModel(labels) {
    const model = this.modelManager.getModel(labels);
    return model;
  }

  /**
   * @returns {Promise}
   */
  getSchemaInfo() {
    return getSchemaInfo(this.options.rest.url, this.options.rest.port, this.auth);
  }

  drop() {
    return this.getSchemaInfo()
      .then(result => {
        const cmds = [];
        for (const label in result) {
          for (const p in result[label]) {
            if (result[label][p].index) {
              cmds.push(new Query(this).dropIndex(label, p));
            }
            if (result[label][p].unique) {
              cmds.push(new Query(this).dropConstraint(label, p, 'unique'));
            }
          }
        }
        cmds.push(new Query(this).match('n').detach('n'));
        return Query.runSequence(cmds);
      });
  }

  /**
   * @param {Object} options
   * @param {Boolean} options.force
   */
  sync(options) {
    debug(`sync start`);
    options = options || {};

    return Promise.resolve()
      .then(() => {
        if (options.force) {
          debug(`cleanup database start`);
          return this.drop();
        }
      })
      .then(() => {
        if (options.force) {
          debug(`cleanup database end`);
        }
        const promises = this.modelManager.all.map(m => m.sync());
        return Promise.all(promises);
      })
      .then(syncResults => {
        const errors = [];
        for (const results of syncResults) {
          for (const result of results) {
            if (result.err) {
              errors.push(result.err);
            }
          }
        }
        return errors;
      })
      .then(result => {
        debug(`sync end`);
        return result;
      })
      .catch(err => {
        debug(`sync end with errors`, err);
        return err;
      });
  }

  /**
   * This method creates a new instance of a Query
   * @return {Query}
   */
  Query() {
    return new Query(this);
  }

  /**
   * This method runs a command with params
   * @param {String} cmd - Cypher statement to execute
   * @param {Object} params - Map with parameters to use in statement
   * @return {Promise} Resolves either with the result or the error message
   */
  run(cmd, params) {
    const session = this.driver.session();
    return session.run(cmd, params)
      .then(result => {
        session.close();
        return result;
      })
      .catch(err => {
        session.close();
        return err;
      });
  }

  /**
   * @returns {UUIDv4}
   */
  static uuid4() {
    return uuid.v4();
  }

  /**
   * @returns {UUIDv1}
   */
  static uuid1() {
    return uuid.v1();
  }
}

Neo4js.prototype.Error = Neo4js.Error = neo4jsErrors.BaseError;
Neo4js.prototype.Model = Neo4js.Model = Model;
Neo4js.prototype.ModelObject = Neo4js.ModelObject = ModelObject;

export default Neo4js;
