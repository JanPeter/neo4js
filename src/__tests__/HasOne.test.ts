import neo4js, {
  Model,
  ModelInstance,
  model,
  relation,
  hasMany,
  hasOne,
  StringProperty,
  NumberProperty,
  HasManyActions,
  HasOneActions,
} from "../index";

type PersonProps = {
  name: string;
};

type TaskProps = {
  title: string;
  done?: boolean;
};

class PersonModel extends Model<PersonProps, PersonInstance> {}
const Person: PersonModel = new PersonModel("Person");

class TaskModel extends Model<TaskProps, TaskInstance> {}
const Task: TaskModel = new TaskModel("Task");

const TaskCreatorRelation = relation
  .from(Person)
  .to(Task)
  .via("created");

const TaskAssigneeRelation = relation
  .from(Person)
  .to(Task)
  .via("assigned");

const SupervisorRelation = relation
  .from(Person)
  .to(Person)
  .via("supervisor");

@model(Person)
class PersonInstance extends ModelInstance<PersonProps> {
  static model = Person;

  @hasMany(() => Task, TaskCreatorRelation)
  createdTasks: HasManyActions<TaskProps, TaskInstance>;

  @hasMany(() => Task, TaskAssigneeRelation)
  assignedTasks: HasManyActions<TaskProps, TaskInstance>;

  @hasMany(() => Person, SupervisorRelation, "in")
  epmloyees: HasManyActions<PersonProps, PersonInstance>;
  @hasOne(() => Person, SupervisorRelation, "out")
  supervisor: HasOneActions<PersonProps, PersonInstance>;
}

@model(Task)
class TaskInstance extends ModelInstance<TaskProps> {
  static model = Task;

  @hasOne(() => Task, TaskCreatorRelation)
  creator: HasOneActions<PersonProps, PersonInstance>;

  @hasOne(() => Task, TaskAssigneeRelation)
  assignee: HasOneActions<PersonProps, PersonInstance>;
}

describe("HasOne", () => {
  beforeAll(() => {
    neo4js.init({
      boltUri: "localhost",
      boltPort: 10001,
    });
  });

  beforeEach(async () => {
    await Person.create({ name: "Olaf" });
    await Person.create({ name: "Ignatz" });
    await Task.create({ title: "Learn magic", done: true });
    await Task.create({ title: "Write more test cases", done: false });
    await Task.create({ title: "Write more test cases", done: false });
  });

  afterEach(async () => {
    await neo4js.run("MATCH (n) DETACH DELETE n");
  });

  afterAll(() => {
    neo4js.close();
  });

  describe("create", () => {
    it("should create one supervisor", async () => {
      const paul: PersonInstance = await Person.create({ name: "Paul" });

      const hubert: PersonInstance = await paul.supervisor.create({
        name: "Hubert",
      });
      delete hubert.props.guid;
      expect(hubert).toMatchSnapshot();
    });

    it("should delete old supervisor when creating new", async () => {
      const paul: PersonInstance = await Person.create({ name: "Paul" });

      const hubert: PersonInstance = await paul.supervisor.create({
        name: "Hubert",
      });
      const olaf: PersonInstance = await paul.supervisor.create({
        name: "Olaf",
      });

      const result = await neo4js.run(
        "MATCH (p:Person)-[:supervisor]->(b:Person) RETURN b"
      );

      expect(result.length).toEqual(1);
      expect(result[0].b).toEqual(olaf.props);
    });
  });

  describe("get", () => {
    it("should get created supervisor", async () => {
      const paul: PersonInstance = await Person.create({ name: "Paul" });

      const hubert: PersonInstance = await paul.supervisor.create({
        name: "Hubert",
      });
      const hubert2: PersonInstance = await paul.supervisor.get();
      expect(hubert.props).toEqual(hubert2.props);
    });

    it("should return null when no relation exists", async () => {
      const paul: PersonInstance = await Person.create({ name: "Paul" });

      const nobody: PersonInstance = await paul.supervisor.get();
      expect(nobody).toEqual(null);
    });
  });

  describe("remove", () => {
    it("should remove supervisor relation without deleting supervisor", async () => {
      const paul: PersonInstance = await Person.create({ name: "Paul" });

      const hubert: PersonInstance = await paul.supervisor.create({
        name: "Hubert",
      });

      let r = await paul.supervisor.remove();
      expect(r).toMatchSnapshot();
      let result: any[] = await neo4js.run("MATCH (m) RETURN m");
      result = result
        .map(node => {
          delete node.m.guid;
          node.m.val = node.m.name ? node.m.name : node.m.title;
          return node;
        })
        .sort((a, b) => a.m.val.localeCompare(b.m.val));
      expect(result).toMatchSnapshot();
    });
  });

  describe("add", () => {
    it("should add created supervisor", async () => {
      const paul: PersonInstance = await Person.create({ name: "Paul" });
      const olaf: PersonInstance = await Person.create({ name: "Olaf" });

      const result = await paul.supervisor.add(olaf);
      expect(result).toMatchSnapshot();

      expect(olaf.props).toEqual((await paul.supervisor.get()).props);
    });
  });

  describe("hasOne", () => {
    it("should return true", async () => {
      const paul: PersonInstance = await Person.create({ name: "Paul" });
      await paul.supervisor.create({ name: "Olaf" });

      const result = await paul.supervisor.hasOne();
      expect(result).toMatchSnapshot();
    });

    it("should return false", async () => {
      const paul: PersonInstance = await Person.create({ name: "Paul" });
      let result = await paul.supervisor.hasOne();
      expect(result).toEqual(false);

      await paul.supervisor.create({ name: "Olaf" });

      result = await paul.supervisor.hasOne();
      expect(result).toEqual(true);

      await paul.supervisor.remove();
      result = await paul.supervisor.hasOne();
      expect(result).toEqual(false);
    });
  });

  describe("update", () => {
    it("should update supervisor", async () => {
      const paul: PersonInstance = await Person.create({ name: "Paul" });
      const olaf: PersonInstance = await paul.supervisor.create({
        name: "Olaf",
      });

      const ignatz: PersonInstance = await paul.supervisor.update({
        name: "Ignatz",
      });
      if (ignatz) {
        expect(olaf.props.guid).toEqual(ignatz.props.guid);
        delete ignatz.props.guid;
        expect(ignatz).toMatchSnapshot();
      } else {
        expect(1).toEqual(0);
      }
    });

    it("should return true if no supervisor can be updated", async () => {
      const paul: PersonInstance = await Person.create({ name: "Paul" });

      const ignatz: PersonInstance = await paul.supervisor.update({
        name: "Ignatz",
      });
      expect(ignatz).toEqual(true);
    });
  });
});
