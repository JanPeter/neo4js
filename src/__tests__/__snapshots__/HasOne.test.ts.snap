// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HasOne add should add created supervisor 1`] = `true`;

exports[`HasOne create should create one supervisor 1`] = `
PersonInstance {
  "assignedTasks": Object {
    "add": [Function],
    "count": [Function],
    "create": [Function],
    "get": [Function],
    "remove": [Function],
    "update": [Function],
  },
  "createdTasks": Object {
    "add": [Function],
    "count": [Function],
    "create": [Function],
    "get": [Function],
    "remove": [Function],
    "update": [Function],
  },
  "epmloyees": Object {
    "add": [Function],
    "count": [Function],
    "create": [Function],
    "get": [Function],
    "remove": [Function],
    "update": [Function],
  },
  "props": Object {
    "name": "Hubert",
  },
  "supervisor": Object {
    "add": [Function],
    "create": [Function],
    "get": [Function],
    "hasOne": [Function],
    "remove": [Function],
    "update": [Function],
  },
}
`;

exports[`HasOne hasOne should return true 1`] = `true`;

exports[`HasOne remove should remove supervisor relation without deleting supervisor 1`] = `
Object {
  "constraintsAdded": 0,
  "constraintsRemoved": 0,
  "indexesAdded": 0,
  "indexesRemoved": 0,
  "labelsAdded": 0,
  "labelsRemoved": 0,
  "nodesCreated": 0,
  "nodesDeleted": 0,
  "propertiesSet": 0,
  "relationshipsCreated": 0,
  "relationshipsDeleted": 1,
}
`;

exports[`HasOne remove should remove supervisor relation without deleting supervisor 2`] = `
Array [
  Object {
    "m": Object {
      "name": "Hubert",
      "val": "Hubert",
    },
  },
  Object {
    "m": Object {
      "name": "Ignatz",
      "val": "Ignatz",
    },
  },
  Object {
    "m": Object {
      "done": true,
      "title": "Learn magic",
      "val": "Learn magic",
    },
  },
  Object {
    "m": Object {
      "name": "Olaf",
      "val": "Olaf",
    },
  },
  Object {
    "m": Object {
      "name": "Paul",
      "val": "Paul",
    },
  },
  Object {
    "m": Object {
      "done": false,
      "title": "Write more test cases",
      "val": "Write more test cases",
    },
  },
  Object {
    "m": Object {
      "done": false,
      "title": "Write more test cases",
      "val": "Write more test cases",
    },
  },
]
`;

exports[`HasOne update should update supervisor 1`] = `
PersonInstance {
  "assignedTasks": Object {
    "add": [Function],
    "count": [Function],
    "create": [Function],
    "get": [Function],
    "remove": [Function],
    "update": [Function],
  },
  "createdTasks": Object {
    "add": [Function],
    "count": [Function],
    "create": [Function],
    "get": [Function],
    "remove": [Function],
    "update": [Function],
  },
  "epmloyees": Object {
    "add": [Function],
    "count": [Function],
    "create": [Function],
    "get": [Function],
    "remove": [Function],
    "update": [Function],
  },
  "props": Object {
    "name": "Ignatz",
  },
  "supervisor": Object {
    "add": [Function],
    "create": [Function],
    "get": [Function],
    "hasOne": [Function],
    "remove": [Function],
    "update": [Function],
  },
}
`;
